design/StableRegions.md

```diff
@@ -0,0 +1,342 @@
+ - The `block-region` table gives a relative position and region ID for each block ID.
+
+Once each regions' vectors have been allocated, the block-region table says how to fill them, one entry at a time.
+Unlike the Rust design, vector entries can be populated out-of-order.
+
+
+### Special (reserved) regions
+
+  - Region 0 -- Anonymous region, for supporting the legacy API that we have today, which lacks `Region` values.
+  - Region 1 -- "Reclaimed blocks" region that consists of blocks reclaimed from GC'd regions.
+  - Regions 2-15 -- Future use by Motoko RTS (TBD).
+
+### Overview of GC support (future work)
+
+- Regions are reprensented (see special subsection) with heap objects that are `stable`, but mutable.
+- They have special GC headers to recognize their special structure.
+- The Region table keeps track of which available IDs are in use as Region heap values are GC'd.
+
+#### Region allocation, re-allocation
+
+A region with a particular ID is first initially allocated by a global, bumped counter reaching it.
+
+When that happens, the region table is marked (TO DO) indicating that region is in use.
+
+The region heap object with that ID witnesses its lifetime, and is reclaimed by GC when it goes out of scope.
+
+When the bump counter reaches its maximum value, the allocation algorithm changes to scanning the regions,
+looking for the first available one (with minimum ID).
+
+#### Region reclamation
+
+A heap object with a certain region ID lives until GC reclaims it, where the region table is re-marked (TO DO) to indicate that the region is free again.
+
+Before reusing that region, each of its block are added to the special "free blocks" region.
+
+
+### Migration from earlier designs into region system
+
+#### Stable memory format versions
+
+The first 32 bits of stable memory record a "marker," which indicates how to determine the "version number"
+for Motoko stable memory, which is stored either:
+  - implicitly, when marker is non-zero, and version is `0`.
+  - explicitly, when marker is zero, and version is stored elsewhere (but currently always `1`).
+
+Including this design, there are three possible verions (`0`, `1`, or `2`):
+
+ 0. Stable vars only. (Here, the marker is the always-non-zero serialization size).
+ 1. Stable vars *plus* direct access to IC0 API, including `grow`.
+ 2. Region system, where direct access still works through region zero.
+
+In the first cases (`0`) and (`1`), we wish to *selectively* migrate into the region system (`2`), with its own internal versioning.
+
+The main factor involved in making this per-canister choice involves
+
+the mandatory block-sized reservation for current and future meta data (8MB).
+
+For canisters that do not store an order of magnitude more than 8MB, it may make sense to
+eschew regions in favor of stable vars for stable data, thereby saving that 8MB of otherwise-wasted space.
+
+##### Explicit, opt-in migration API
+
```

Comment:
@crusso I wrote this little section of the doc today to download my current thoughts, in light of the concerns about the 8MB minimum if we don't have an "opt in" mechanism.

To me, it represents the steady-state API of the compiler, and also what I hope to merge in this PR.  Basically, because of the 8MB minimum for regions, we should plan to keep it behind an "opt in" API forever.

If it seems reasonable, then I think I can take a stab at implementing the flag mechanism next.